Strassen Matrix Multiplication/Matrix.swift

```diff
@@ -0,0 +1,246 @@
+//
+//  Matrix.swift
+//  
+//
+//  Created by Richard Ash on 10/28/16.
+//
+//
+
+import Foundation
+
+enum RowOrColumn {
+  case row, column
+}
+
+struct Matrix<T: Number> {
+  
+  // MARK: - Variables
+  
+  let rows: Int, columns: Int
+  var grid: [T]
+  
+  var isSquare: Bool {
+    return rows == columns
+  }
+  
+  var size: MatrixSize {
+    return MatrixSize(rows: rows, columns: columns)
+  }
+  
+  // MARK: - Init
+  
+  init(rows: Int, columns: Int, initialValue: T = T.zero) {
+    self.rows = rows
+    self.columns = columns
+    self.grid = Array(repeating: initialValue, count: rows * columns)
+  }
+  
+  init(size: Int, initialValue: T = T.zero) {
+    self.rows = size
+    self.columns = size
+    self.grid = Array(repeating: initialValue, count: rows * columns)
+  }
+  
+  // MARK: - Subscript
+  
+  subscript(row: Int, column: Int) -> T {
+    get {
+      assert(indexIsValid(row: row, column: column), "Index out of range")
+      return grid[(row * columns) + column]
+    }
+    set {
+      assert(indexIsValid(row: row, column: column), "Index out of range")
+      grid[(row * columns) + column] = newValue
+    }
+  }
+  
+  subscript(type: RowOrColumn, value: Int) -> [T] {
+    get {
+      switch type {
+      case .row:
+        assert(indexIsValid(row: value, column: 0), "Index out of range")
+        return Array(grid[(value * columns)..<(value * columns) + columns])
+      case .column:
+        assert(indexIsValid(row: 0, column: value), "Index out of range")
+        var columns: [T] = []
+        for row in 0..<rows {
+          for column in 0..<self.columns {
+            if column == value {
+              columns.append(grid[(row * self.columns) + value])
+            }
+          }
+        }
+        return columns
+      }
+    } set {
+      switch type {
+      case .row:
+        assert(newValue.count == columns)
+        for (column, element) in newValue.enumerated() {
+          grid[(value * columns) + column] = element
+        }
+      case .column:
+        assert(newValue.count == rows)
+        for (row, element) in newValue.enumerated() {
+          grid[(row * columns) + value] = element
+        }
+      }
+    }
+  }
+  
+  // MARK: - Private Functions
```

Comment:
Consider putting all the `private` methods in a `private` extension. It's pretty nice in that it visually groups the helper methods together, and has a few architectural benefits as well - nobody would be able to put a differently scoped method / computed property in the middle of your `private` section.